Source/ThirdParty/ANGLE/src/tests/gl_tests/VariableRasterizationRateTestMetal.mm

```diff
@@ -0,0 +1,300 @@
+//
+// Copyright 2021 The ANGLE Project Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+//
+// ImageTestMetal:
+//   Tests the correctness of eglImage with native Metal texture extensions.
+//
+
+#include "test_utils/ANGLETest.h"
+
+#include "common/mathutil.h"
+#include "test_utils/gl_raii.h"
+#include "util/EGLWindow.h"
+
+#include <CoreFoundation/CoreFoundation.h>
+#include <Metal/Metal.h>
+
+namespace angle
+{
+namespace
+{
+constexpr int kTileSize                        = 32;
+constexpr int kWindowWidth                     = 8 * kTileSize;
+constexpr int kWindowHeight                    = kWindowWidth;
+constexpr int kZones                           = 7;
+constexpr float kQuality[kZones]               = {0.24f, 0.24f, 0.24f, 0.24f, 0.49f, 0.49f, 0.49f};
+constexpr int kViewportWidth                   = 3 * kTileSize;
+constexpr int kViewportHeight                  = 3 * kTileSize;
+constexpr char kVariableRasterizationRateExt[] = "GL_ANGLE_variable_rasterization_rate_metal";
+constexpr char kVS[]                           = R"(precision highp float;
+attribute vec4 position;
+varying vec2 texcoord;
+
+void main()
+{
+    gl_Position = position;
+    texcoord = (position.xy * 0.5) + 0.5;
+    texcoord.y = 1.0 - texcoord.y;
+})";
+constexpr char kFS[]                           = R"(
+precision highp float;
+varying vec2 texcoord;
+
+void main()
+{
+    gl_FragColor = vec4(texcoord, 0.0, 1.0);
+})";
+constexpr char kCheckerBoardFS[]               = R"(
+precision highp float;
+uniform vec2 u_resolution;
+uniform float u_checkCount;
+
+void main()
+{
+    vec2 uv = gl_FragCoord.xy / u_resolution.xy;
+    uv.x *= u_resolution.x / u_resolution.y;
+    uv *= u_checkCount;
+
+    float u = 1.0 - floor(mod(uv.x, 2.0));
+    float v = 1.0 - floor(mod(uv.y, 2.0));
+    vec3 col = vec3(1.0);
+
+    if(u == 1.0 && v < 1.0 || v == 1.0 && u < 1.0) {
+        col = vec3(0.);
+    }
+    gl_FragColor = vec4(col,1.0);
+})";
+}  // anonymous namespace
+
+class ScopeMetalRasterizationRateMapRef : angle::NonCopyable
+{
+  public:
+    explicit ScopeMetalRasterizationRateMapRef(id<MTLRasterizationRateMap> &&map) : mMap(map) {}
+
+    ScopeMetalRasterizationRateMapRef(const ScopeMetalRasterizationRateMapRef &other)
+    {
+        if (mMap)
```

Comment:
This is a copy constructor, so `this` is starting in a clean state - how would `mMap` be non-nil?

Did you mean to define `operator=`? If so and if you were aiming to properly manage refcounts, I think it's important to ref the incoming `other.mMap` before unref'ing `this`'s in order to handle self-assignment. Not 100% sure.